blog_article/blog_article.adoc

```diff
@@ -0,0 +1,161 @@
+used with partially limited functionality.
+
+== AMQP 0.9.1 and AMQP 1.0 Interoperability
+As we already know, AMQP 0.9.1 and AMQP 1.0 should be considered as completely different specifications which leads
+to the questions: Is interoperability between these two protocols possible
+"out-of-the-box"? And if not, what has to be considered to achieve interoperability? What about interoperability
+between different clients speaking the same protocol? +
+These points are especially important when we think again about an (evolving) event-driven architecture. The components
+of a system should be able to exchange events and data independently of the protocol they are using. +
+As also described within RabbitMQ’s plugin documentation, interoperability between AMQP 0.9.1 and AMQP 1.0 is provided
+with some constraints:
+
+|===
+|Producer Library |Consumer Library|Interoperability
+
+|SwiftMQ 1.0
+|RabbitMQ 0.9.1
+|provided with constraints -> sending as plain bytes
+
+|RabbitMQ 0.9.1
+|SwiftMQ 1.0
+|provided with constraints -> reading from data field
+
+|Qpid JMS 1.0
+|SwiftMQ 1.0
+|provided
+
+|SwiftMQ 1.0
+|Qpid JMS 1.0
+|provided
+
+|Qpid JMS 1.0
+|RabbitMQ 0.9.1
+|provided with constraints -> sending as plain bytes
+
+|RabbitMQ 0.9.1
+|Qpid JMS 1.0
+|provided
+|===
+
+Like denoted in the table above, when testing the different clients, it was generally possible to send and receive
+messages using AMQP 1.0 and AMQP 0.9.1. +
+For AMQP 1.0 messages that should be read by AMQP 0.9.1 clients, it should be considered that AMQP 1.0 uses
+the so-called AMQP-encoding. This means that all messages that are sent with AMQP 1.0 are encoded beforehand.
+The grammar of this encoding is specified within the AMQP 1.0 protocol. However, this grammar can be implemented
+differently by each client. +
+The consequences will become clear when looking at an example:
+
+* Sent message using SwiftMQ client (AMQP 1.0): M1
+* Received message using RabbitMQ client (AMQP 0.9.1): �w�M1
+
+* Sent message using Qpid JMS client (AMQP 1.0): M1
+* Received message using RabbitMQ client (AMQP 0.9.1): Sw�M1
+
+The examples show that there are some additional bytes appended in front of the messages. The reason for this is the
+AMQP-encoding that is used by the message producers. Since this encoding seems not to be implemented the same for
+every client, the AMQP-plugin cannot cut those additional bytes off. This issue was also discussed in the RabbitMQ
+https://groups.google.com/g/rabbitmq-users/c/QBsc_wCuTBs/m/xh84gj1hAwAJ[mailing list].
+Despite the slightly different implementation of the encoding and decoding of the clients, it is possible to send
+and receive AMQP 1.0 messages with different AMQP 1.0 client libraries without constraints.
+
+The AMQP 1.0 protocol also includes a possibility to send unencoded messages. Data can be sent as payload of the
+_data field_ of an AMQP 1.0 message. Data included in this _data field_ will not be encoded and therefore sent as plain
```

Comment:
Basically, if binary is sent via data section, it is also encoded! But as "data", and therefore, the RabbitMQ plugin can handle it. The plugin also only handles it if data is the first section. Amqp allows it to to add multiple data sections, and also to add binary data as part of an amqp-value or an ampq-sequence. More than one data sections as well binary as port of amqp-value (what is if binary is the only value of an amqp-value) cannot be handled, because AMQP 0.9.1 only allows a single data value and therefore the plugin cannot knwo how to translate it.

As described here (I know you know, but I would mention this), the body (application-data) can consist of one ore more data sections, one or more amqp-sequence sections or a single-value section: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format
A data section is defined as: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-data
Basically data is a "described type": http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#doc-idp38080
So the actual binary content is annotated with a descriptor, that states that this is "data". This is exactly what is transferred over the wire if "M1" is sent as binary data (UTF8):

        Constructor         Untyped Bytes
+---------------------+----------------+
|                                 |                         |
0x00 0x53 0x75 0xA0 0x02 0x4D 0x32
        |                |        |    |      M      1
        +---------+-----+   |  
        Descriptor    |         |
          for data      |      bytes of binary data
                             |
                 primitive format code for
                 2⁸-1 length binary (http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-binary)